src/routes/clients.js

```diff
@@ -31,13 +33,26 @@ let router = express.Router();
 // Create a new API client.
 router.post('/', (req, res) => {
   // Required params: name
-  // XXX We will probably require redirect url for signin/signup flow.
   req.checkBody('name', 'missing or invalid required "name" parameter')
      .notEmpty();
 
-  const errors = req.validationErrors();
-  if (errors) {
-    return ApiError(res, 400, ERRNO_INVALID_API_CLIENT_NAME, BAD_REQUEST);
+  if (req.body.redirectUrl) {
+    req.checkBody('redirectUrl', 'invalid "redirectUrl"')
+       .isUrl({ require_valid_protocol: true });
+  }
+
+  const error = req.validationErrors()[0];
+  if (error) {
+    let errno = ERRNO_BAD_REQUEST;
+    switch (error.param) {
+      case 'redirectUrl':
+        errno = ERRNO_INVALID_API_CLIENT_REDIRECT_URL;
+        break;
+      case 'name':
+        errno = ERRNO_INVALID_API_CLIENT_NAME;
+        break;
+    }
+    return ApiError(res, 400, errno, BAD_REQUEST);
```

Comment:
(not for this patch either)

I don't really like how the errors are generated currently. For example I don't get why we need to put the numeric error number when errors.js already have it... It's also difficult to remember the order of the parameters.

Here is how I did it in a previous app:

* in the main app configuration file: https://github.com/project-abigail/calendar-server/blob/master/app.js#L46-L68
* here are my errors: https://github.com/project-abigail/calendar-server/blob/master/utils/errors.js
* here is an example about how to use them: https://github.com/project-abigail/calendar-server/blob/master/routes/users.js#L38-L41